src/auction.js

```diff
@@ -353,19 +188,128 @@ export function newAuction({adUnits, adUnitCodes, callback, cbTimeout}) {
     });
 
     _auctionStatus = AUCTION_IN_PROGRESS;
-    adaptermanager.callBids(_adUnits, bidRequests, addBidResponseDecorator(addBidResponse.bind(this)), done.bind(this));
+    adaptermanager.callBids(_adUnits, bidRequests, addBidResponse.bind(this), done.bind(this));
   };
 
   return {
+    addBidReceived,
+    executeCallback,
+    callBids,
+    getTimeout: () => _timeout,
+    getAuctionId: () => _auctionId,
     getAuctionStatus: () => _auctionStatus,
     getAdUnits: () => _adUnits,
     getAdUnitCodes: () => _adUnitCodes,
     getBidRequests: () => _bidderRequests,
     getBidsReceived: () => _bidsReceived,
-    callBids
   }
 }
 
+export const addBidResponse = createHook('asyncSeries', function addBidResponse(adUnitCode, bid) {
+  let auctionInstance = this;
+  let bidRequests = auctionInstance.getBidRequests();
+  let auctionId = auctionInstance.getAuctionId();
+
+  let bidResponse = getPreparedBidForAuction({adUnitCode, bid, bidRequests, auctionId});
+  if (bidResponse.mediaType === 'video') {
+    tryAddVideoBid(bidResponse);
+  } else {
+    doCallbacksIfNeeded();
+    addBidToAuction(bidResponse);
+  }
+
+  function doCallbacksIfNeeded() {
+    if (bidResponse.timeToRespond > auctionInstance.getTimeout() + config.getConfig('timeoutBuffer')) {
+      auctionInstance.executeCallback(true);
+    }
+  }
+
+  // Add a bid to the auction.
+  function addBidToAuction() {
+    events.emit(CONSTANTS.EVENTS.BID_RESPONSE, bidResponse);
+    auctionInstance.addBidReceived(bidResponse);
+  }
+
+  // Video bids may fail if the cache is down, or there's trouble on the network.
+  function tryAddVideoBid(bidResponse) {
+    if (config.getConfig('usePrebidCache')) {
+      store([bidResponse], function(error, cacheIds) {
+        if (error) {
+          utils.logWarn(`Failed to save to the video cache: ${error}. Video bid must be discarded.`);
+        } else {
+          bidResponse.videoCacheKey = cacheIds[0].uuid;
+          if (!bid.vastUrl) {
+            bidResponse.vastUrl = getCacheUrl(bidResponse.videoCacheKey);
+          }
+          addBidToAuction(bidResponse);
+        }
+        doCallbacksIfNeeded();
+      });
+    } else {
+      doCallbacksIfNeeded();
+      addBidToAuction(bidResponse);
+    }
+  }
+});
+
+// Postprocess the bids so that all the universal properties exist, no matter which bidder they came from.
+// This should be called before addBidToAuction().
+function getPreparedBidForAuction({adUnitCode, bid, bidRequests, auctionId}) {
```

Comment:
How come this was moved outside `addBidResponse`?

Reply:
addBidResponse function had multiple inner functions like `isValid`, `preparedBidForAuction` and `validBidSize` which was creating confusing so moved validation logic to `bidderFactory` and also moved `getPreparedBidForAuction`. 

Previously `preparedBidForAuction` was mutating bid object and now its not.